### PR TITLE
fix: add lock timeout to reduce lock failure cases

### PR DESCRIPTION
**Pull request by allencloud**
Signed-off-by: Allen Sun <allensun.shl@alibaba-inc.com>

### Ⅰ. Describe what this PR did

`failed to xxx container xxxx: lock failed` is a normal error when pouchd meets lots of concurrent request for the same container object, such as `container`.

In the original implementation, ctrd will try to get the lock for one container to continue the request's operation. While if the container's access(lock) is owned by others, then the request will not be blocked and will return a lock failure instead. This action has benefit and shortcomings as well. The benefit is that it could increase the real time of request even if it fails. The drawback is that it would depends on the client callers to retry the request.

To avoid this, we could take other ways like:

* make the lock infinite blocked;
* add a timeout when accessing the lock.

This pull request uses the second way to make it.

Besides the logic update for timeout(default is 5 seconds). I also changed the following things:

* make the lock failure error to be more explicit, like `container xxx is accessed by other request and please try again: lock failed`;
* remove the containerLock in ctrd since it could be replaced by kmutex pkg.

### Ⅱ. Does this pull request fix one issue?

fix https://github.com/alibaba/pouch/issues/2281

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
no need for this, I added more test cases for the common package knutex.

### Ⅳ. Describe how to verify it
none

### Ⅴ. Special notes for reviews
none

### CODE REVIEW

File: pkg/kmutex/kmutex.go [hunk 1]
```diff
@@ -119,12 +139,15 @@ func (m *KMutex) Lock(k string) bool {
...
 func (m *KMutex) Unlock(k string) {
 	m.Mutex.Lock()
 	defer m.Mutex.Unlock()
 
-	v, ok := m.keys[k]
+	v, ok := m.lockedKeys[k]
 	if ok && len(v.c) == 0 {
+		// filled up the chan to identify caller has released the lock.
 		v.c <- struct{}{}
 	}
```
Comment (xiaoxubeii) [inline pkg/kmutex/kmutex.go]: Should it free key when not used anymore. Memory will increase when more and more keys are created. Some codes like:

```
if v.waits == 0 {
		// Free the lock if no one is holding or trying to hold.
		delete(m.lockedKeys, k)
	}
```

File: pkg/kmutex/kmutex.go [hunk 2]
```diff
@@ -119,12 +139,15 @@ func (m *KMutex) Lock(k string) bool {
...
 }
```
Comment (xiaoxubeii) [inline pkg/kmutex/kmutex.go]: Should it free key when not used anymore. Memory will increase when more and more keys are created. Some codes like:

```
if v.waits == 0 {
		// Free the lock if no one is holding or trying to hold.
		delete(m.lockedKeys, k)
	}
```

Comment (xiaoxubeii) [inline pkg/kmutex/kmutex.go]: Is it better that free unused lockedKeys in this `Unlock` method than using a extra go routine?
https://github.com/alibaba/pouch/blob/f56461421b3a2148db94d825a5e715fb156e2972/pkg/kmutex/kmutex.go#L41-L52

Comment (allencloud) [reply pkg/kmutex/kmutex.go]: Actually this is a legacy function which is designed by @skoo87 .
Could you help explain a little bit more about the goroutine usage here?

File: pkg/kmutex/kmutex.go [hunk 3]
```diff
@@ -68,15 +81,19 @@ func (m *KMutex) Trylock(k string) bool {
...
 	default:
+		// return false immediately if someone locked it.
 		return false
 	}
 }
 
-// LockWithTimeout trys to lock, if can't acquire the lock, will block util timeout.
+// LockWithTimeout tries to lock.
+// It can't acquire the lock, will block util timeout.
 func (m *KMutex) LockWithTimeout(k string, to time.Duration) bool {
```
Comment (allencloud) [inline pkg/kmutex/kmutex.go]: s/LockWithTimeout/TrylockWithTimeout

File: pkg/kmutex/kmutex.go [hunk 4]
```diff
@@ -68,15 +81,19 @@ func (m *KMutex) Trylock(k string) bool {
 		return true
 	}
 
+	// the k has already been locked.
 	select {
 	case <-v.c:
+		// the locker has released the lock.
 		return true
```
Comment (allencloud) [inline pkg/kmutex/kmutex.go]: In this line, we think there is a bug there.
If this goroutine gets the lock and returns true, there is a very tiny possibility that the GC function would delete the corresponding locked key.

Comment (allencloud) [inline pkg/kmutex/kmutex.go]: In this line, we think there is a bug there.
If this goroutine gets the lock and returns true, there is a very tiny possibility that the GC function would delete the corresponding locked key.

**Comment by allencloud** [comment]
Currently block this pull request. For the issue #2281 , we will try to make the code : 
```
if !c.Trylock(id) {
		return nil, errtypes.ErrLockfailed			
}		
```

to be:
```
if !c.Trylock(id) {
      for i:= 0;i<3;i++{
            // try another 3 times.
     }			
}
```

File: pkg/kmutex/kmutex.go (hunk 4 above)
Comment (xiaoxubeii) [reply pkg/kmutex/kmutex.go]: Agree. These lock functions, such as `Trylock`, `LockWithTimeout ` and `Lock` all have same problem that they didn't not get the lock after `<-v.c`.

**Comment by fuweid** [comment]
this issue has been fixed by https://github.com/alibaba/pouch/pull/2319

**Pull request closed without merging**